src/wallet/walletdb.cpp

```diff
@@ -1227,6 +1232,21 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         result = DBErrors::CORRUPT;
     }
 
+    // Since it was accidentally possible to "encrypt" a wallet with private keys disabled, we should check if this is
+    // such a wallet and remove the encryption key records to avoid any future issues.
+    // Although wallets without private keys should not have *ckey records, we should double check that.
+    // Removing the mkey records is only safe if there are no *ckey records.
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && pwallet->HasEncryptionKeys() && !pwallet->HaveCryptedKeys()) {
```

Comment:
I notice that nothing in this block requires any locks (even though `cs_wallet` will be locked at this stage in this function).

For my own understanding, is this because the db operations, i.e. writing to the on-disk file, have their synchronisation handled by the db's transaction system?

I see for example in `LoadEncryptionKey` that we lock `cs_wallet` to load the keys _into_ the wallet, which makes intuitive sense. But it feels that conversely there should be something asserting the lock is held here too, as we remove encryption keys from the wallet? Or at least when `pwallet->mapMasterKeys.clear();` is called.

Perhaps it doesn't matter as it's done during loading from the db, so nothing else will have started to access the Wallet?

Reply:
The lock is already being held from the top of the function.

Reply:
I don’t know where else to go for help, but this sounds like exactly what I’ve been looking for (hopefully). About a year ago I found my old pre-2017 BTC info & when viewing it I accidentally imported it as a WatchOnly & now I can’t do anything with it. I’m a Ethereum Validator & never really understood the BTC Ecosystem, so any help in the right direction would be appreciated. I’ll donate 0.02 BTC to whoever can help.